internal/topology/variables/patch_validation.go

```diff
@@ -0,0 +1,222 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package variables
+
+import (
+	"fmt"
+	"log"
+	"strings"
+	"text/template"
+
+	"github.com/dolmen-go/jsonptr"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+// ValidatePatches returns errors if the Patches in the ClusterClass violate any validation rules.
+func ValidatePatches(patches []clusterv1.ClusterClassPatch, variables []clusterv1.ClusterClassVariable, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	names := sets.String{}
+	for i, patch := range patches {
+		allErrs = append(
+			allErrs,
+			validatePatch(patch, names, variables, path.Index(i))...,
+		)
+		names.Insert(patch.Name)
+	}
+	return allErrs
+}
+
+func validatePatch(patch clusterv1.ClusterClassPatch, names sets.String, variables []clusterv1.ClusterClassVariable, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	allErrs = append(allErrs,
+		validatePatchName(patch, names, path)...,
+	)
+	log.Print(allErrs.ToAggregate())
+	allErrs = append(allErrs,
+		validatePatchDefinitions(patch, variables, path)...,
+	)
+	return allErrs
+}
+
+func validatePatchName(patch clusterv1.ClusterClassPatch, names sets.String, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if patch.Name == "" {
+		allErrs = append(allErrs,
+			field.Invalid(
+				path.Child("name"),
+				patch.Name,
+				"patch names must not be empty",
+			),
+		)
+	}
+
+	if names.Has(patch.Name) {
+		allErrs = append(allErrs,
+			field.Invalid(
+				path.Child("name"),
+				patch.Name,
+				fmt.Sprintf("patch names must be unique. %s is repeated.", patch.Name),
+			),
+		)
+	}
+	return allErrs
+}
+
+/*
+
+
+- We consider validating the Go template by generating values for all variables used in the template, then evaluating
+the template and validating if the result is a valid YAML or JSON value.
+
+*/
+func validatePatchDefinitions(patch clusterv1.ClusterClassPatch, variables []clusterv1.ClusterClassVariable, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	for _, definition := range patch.Definitions {
+		allErrs = append(allErrs,
+			validatePatchDefinitionSelector(definition.Selector, path)...)
+
+		allErrs = append(allErrs,
+			validatePatchJSONPatches(definition.JSONPatches, variables, path)...)
+	}
+	return allErrs
+}
+
+func validatePatchJSONPatches(jsonPatches []clusterv1.JSONPatch, variables []clusterv1.ClusterClassVariable, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	validOps := sets.NewString("add", "replace", "remove")
+	for i, jsonPatch := range jsonPatches {
+		if !validOps.Has(jsonPatch.Op) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Child("op").Index(i),
+					jsonPatch,
+					fmt.Sprintf("jsonPatch op must be one of \"add\" \"remove\" or \"replace\". %v is not valid.", jsonPatch.Op),
```

Comment:
Should this be
```suggestion
					path.Index(i).Child("op"),
					jsonPatch.Op,
					fmt.Sprintf("jsonPatch op must be one of \"add\" \"remove\" or \"replace\""),
```
Same for the other checks in the loop

Reply:
Can we use `field.NotSupported` in this case? (we did the same in the clusterclass variable validation and it allows us to avoid duplicating the list of valid values)